overlord/devicestate/devicestate.go

```diff
@@ -341,91 +344,270 @@ func extractDownloadInstallEdgesFromTs(ts *state.TaskSet) (firstDl, lastDl, firs
+
+	// new model specifies a different snap
+	needsInstall, err := notInstalled(st, ms.newModelSnap.SnapName())
+	if err != nil {
+		return nil, err
+	}
+	if needsInstall {
+		// which needs to be installed
+		return snapstateInstallWithDeviceContext(ctx, st, ms.newSnap,
+			&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+			userID, snapstate.Flags{}, deviceCtx, fromChange)
 	}
-	if current.Base() != new.Base() {
-		needsInstall, err := notInstalled(st, new.Base())
+
+	if ms.new.Grade() != asserts.ModelGradeUnset {
+		// pre UC20 remodels would not update an already installed snap
+
+		// which is already installed, but may be using a different channel
+		changed, err := installedSnapChange(st, ms.newModelSnap.SnapName(), newModelSnapChannel)
 		if err != nil {
 			return nil, err
 		}
-		if needsInstall {
-			ts, err = snapstateInstallWithDeviceContext(ctx, st, new.Base(), nil, userID, snapstate.Flags{}, deviceCtx, fromChange)
+		if changed {
+			ts, err := snapstateUpdateWithDeviceContext(st, ms.newSnap,
+				&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
+			if err != nil {
+				return nil, err
+			}
+			if ts != nil {
+				if edgeTask := ts.MaybeEdge(snapstate.DownloadAndChecksDoneEdge); edgeTask != nil {
+					// we have downloads and checks done edge, so
+					// the update is not a simple
+					// switch-snap-channel
+					return ts, nil
+				} else {
+					// in other cases make sure that the
+					// kernel or base is linked and available
+					return snapstate.AddLinkNewBaseOrKernel(st, ts)
+				}
+			}
+		}
+	}
+	return snapstate.LinkNewBaseOrKernel(st, ms.newSnap)
+}
+
+func remodelGadgetTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+	userID := 0
+	newGadgetChannel, err := modelSnapChannelFromDefaultOrPinnedTrack(ms.new, ms.newModelSnap)
+	if err != nil {
+		return nil, err
+	}
+	if ms.currentSnap == ms.newSnap {
+		// which is already installed, but may be using a different channel
+		changed := false
+		if ms.new.Grade() != asserts.ModelGradeUnset {
+			// UC20 models can specify default channel for all snaps
+			// including the gadget
+			changed, err = installedSnapChange(st, ms.newSnap, newGadgetChannel)
+			if err != nil {
+				return nil, err
+			}
 		} else {
-			ts, err = snapstate.LinkNewBaseOrKernel(st, new.Base())
+			// pre UC20 models could only specify a track for the
+			// gadget
+			changed = ms.currentModelSnap.PinnedTrack != ms.newModelSnap.PinnedTrack
 		}
-		if err != nil {
-			return nil, err
+		if changed {
+			return snapstateUpdateWithDeviceContext(st, ms.newSnap,
+				&snapstate.RevisionOptions{Channel: newGadgetChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
 		}
+		return nil, nil
+	}
+
+	// install the new gadget
+	return snapstateInstallWithDeviceContext(ctx, st, ms.newSnap,
```

Comment:
The system can have only one gadget, so the scenario when a different gadget is already installed is not possible.

Reply:
Hm, so I think there's a problem here. Per quick discussion with @pedronis we may have been doing the wrong thing and missed the step about gadget assets update. Although the scenario needed here is to have the old gadget snap (A), then remodel to new gadget snap (B) with new boot assets, and remodel again back to (A). FWIW I don't think it's a regression introduced in this PR.

Reply:
I agree it's not a regression but that scenario you describe is one that I think is important for our remodelling story since that's what a "white box OEM" wants to enable, so they can collect already branded devices and unbrand them, back to the generic box so that another end-user/end-customer could brand it to some other brand, which would have 3 remodels:
1. remodel from generic brand to specific brand
2. remodel back to generic brand when done being a specific brand
3. remodel to a new different specific brand when resold
4. remodel back to generic brand when done being a specific brand again 
and so and so forth